src/app/modules/submission/email-submission/email-submission.util.ts

```diff
@@ -535,18 +541,172 @@ export const maskUidOnLastField = (
+    const checkbox = getAnswerForCheckbox(response)
+    return [getFormattedFunction(checkbox, hashedFields)]
+  } else {
+    return [getFormattedFunction(response, hashedFields)]
+  }
+}
+
+/**
+ * Helper function to mask NRICs in Corppass Validated UID
+ */
+const maskField = (field: EmailAutoReplyField): EmailAutoReplyField => {
+  const maskedAnswerTemplate = field.answerTemplate.map((answer) => {
+    return answer.length >= 4 // defensive, in case UID length is less than 4
+      ? '*'.repeat(answer.length - 4) + answer.substr(-4)
+      : answer
+  })
+  return {
+    question: field.question,
+    answerTemplate: maskedAnswerTemplate,
+  }
+}
+
+/**
+ * Function to extract information for email json field from response
+ * Json field is used for data collation tool
+ */
+const getJsonFormattedResponse = (
+  response: ResponseFormattedForEmail,
+): EmailJsonField | undefined => {
+  const { answer, fieldType } = response
+  // Headers are excluded from JSON data
+  if (fieldType !== BasicField.Section) {
+    const jsonData = {
+      question: getJsonPrefixedQuestion(response),
+      answer,
+    }
+    return jsonData
+  }
+  return undefined
+}
+
+/**
+ * Function to extract information for email form field from response
+ * Form field is used to send responses to admin
+ */
+const getFormFormattedResponse = (
+  response: ResponseFormattedForEmail,
+  hashedFields: Set<string>,
+): EmailFormField => {
+  const { answer, fieldType } = response
+  const answerSplitByNewLine = answer.split('\n')
+  const formData = {
+    question: getFormDataPrefixedQuestion(response, hashedFields),
+    answerTemplate: answerSplitByNewLine,
+    answer,
+    fieldType,
+  }
+  return formData
+}
+
+/**
+ * Function to extract information for email autoreply field from response
+ * Autoreply field is used to send confirmation emails
+ */
+const getAutoReplyFormattedResponse = (
+  response: ResponseFormattedForEmail,
+): EmailAutoReplyField | undefined => {
+  const { question, answer, isVisible } = response
+  const answerSplitByNewLine = answer.split('\n')
+  // Auto reply email will contain only visible fields
+  if (isVisible) {
+    const autoReplyData = {
+      question, // No prefixes for autoreply
+      answerTemplate: answerSplitByNewLine,
+    }
+    return autoReplyData
+  }
+  return undefined
+}
+
+export class EmailDataObj {
+  parsedResponses: ProcessedFieldResponse[]
+  hashedFields: Set<string>
+  authType: AuthType
+
+  constructor(
+    parsedResponses: ProcessedFieldResponse[],
+    hashedFields: Set<string>,
+    authType: AuthType,
+  ) {
+    this.parsedResponses = parsedResponses
+    this.hashedFields = hashedFields
+    this.authType = authType
+  }
+
+  /**
+   * Getter function to return jsonData which is used for data collation tool
+   */
+  get jsonData(): EmailJsonField[] {
+    return this.parsedResponses.flatMap((response) =>
+      compact(
+        createFormattedDataForOneField(
```

Comment:
would it be better to separate these nested functions into different lines, and add comments explaining why the `compact` necessary? can also consider moving the `compact` to the final return statement so it's only called once. having 3 nested function calls makes this hard to understand

Reply:
edited, see if this works